src/lib/mina_lib/mina_lib.ml

```diff
@@ -778,10 +780,89 @@ let last_epoch_delegators t ~pk =
   in
   find_delegators last_epoch_delegatee_table pk
 
+let perform_compaction t =
+  if
+    not
+      (Genesis_constants.Proof_level.equal
+         t.config.precomputed_values.proof_level Full)
+  then ()
+  else
+    let slot_duration_ms =
+      let leeway = 1000 in
+      t.config.precomputed_values.constraint_constants.block_window_duration_ms
+      + leeway
+    in
+    let expected_time_for_compaction =
+      match Sys.getenv "MINA_COMPACTION_MS" with
+      | Some ms ->
+          Float.of_string ms
+      | None ->
+          6000.
+    in
+    let span ?(incr = 0.) ms = Float.(of_int ms +. incr) |> Time.Span.of_ms in
+    let interval_configured =
+      match Sys.getenv "MINA_COMPACTION_INTERVAL_MS" with
+      | Some ms ->
+          Time.Span.of_ms (Float.of_string ms)
+      | None ->
+          span (slot_duration_ms * 2)
+    in
+    if Time.Span.(interval_configured <= of_ms expected_time_for_compaction)
+    then (
+      [%log' fatal t.config.logger]
+        "Time between compactions %f should be greater than the expected time \
+         for compaction %f"
+        (Time.Span.to_ms interval_configured)
+        expected_time_for_compaction ;
+      failwith
```

Comment:
Does this need to be fatal? It appears that below we don't start the next `upon` timer until the previous compaction has run, so I don't think it's an existential problem (although it is unfortunate to spend >50% of compute time doing GC).

Reply:
> although it is unfortunate to spend >50% of compute time doing GC.

Added the failure mainly for this. Compactions should be sufficiently spaced otherwise it might hinder normal node operations